File: evidence-service/src/response-cache.ts
/**
 * In-memory response cache with per-type TTLs.
 *
 * Eliminates redundant upstream Tradier calls within short windows.
 * Process-local only. Resets on server restart. No manual purging needed.
 *
 * TTL policy:
 *   - Expirations: 5 minutes (changes infrequently within a session)
 *   - Quotes: 60 seconds (underlying price, changes slowly relative to options)
 *   - Chains: 90 seconds (prevent duplicate fetches within acquisition pass)
 */

export type CacheType = "expirations" | "quote" | "chain";

interface CacheEntry<T> {
  data: T;
  expiresAt: number;
  retrievedAt: string;
}

const TTL_MS: Record<CacheType, number> = {
  expirations: 5 * 60 * 1000,  // 5 minutes
  quote: 60 * 1000,             // 60 seconds
  chain: 90 * 1000,             // 90 seconds
};

export class ResponseCache {
  private store = new Map<string, CacheEntry<unknown>>();

  /**
   * Get cached response. Returns null on miss or expiry.
   */
  get<T>(type: CacheType, key: string): { data: T; retrievedAt: string } | null {
    const fullKey = `${type}:${key}`;
    const entry = this.store.get(fullKey);
    if (!entry) return null;
    if (Date.now() > entry.expiresAt) {
      this.store.delete(fullKey);
      return null;
    }
    return { data: entry.data as T, retrievedAt: entry.retrievedAt };
  }

  /**
   * Store a response with type-appropriate TTL.
   */
  set<T>(type: CacheType, key: string, data: T, retrievedAt: string): void {
    const fullKey = `${type}:${key}`;
    this.store.set(fullKey, {
      data,
      expiresAt: Date.now() + TTL_MS[type],
      retrievedAt,
    });
  }

  /**
   * Cache statistics for diagnostics.
   */
  stats(): { size: number; byType: Record<CacheType, number> } {
    const byType: Record<CacheType, number> = { expirations: 0, quote: 0, chain: 0 };
    for (const key of this.store.keys()) {
      const type = key.split(":")[0] as CacheType;
      if (byType[type] !== undefined) byType[type]++;
    }
    return { size: this.store.size, byType };
  }
}

// Singleton
let instance: ResponseCache | null = null;

export function getResponseCache(): ResponseCache {
  if (!instance) instance = new ResponseCache();
  return instance;
}

File: evidence-service-java/src/main/java/com/wheelwright/evidence/provider/ResponseCache.java
package com.wheelwright.evidence.provider;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * In-memory response cache with per-type TTLs.
 *
 * Behavioral parity with TypeScript ResponseCache:
 *   - Expirations: 5 minutes
 *   - Quotes: 60 seconds
 *   - Chains: 90 seconds
 *   - Returns null on miss or expiry
 *   - Process-local, resets on restart
 */
public class ResponseCache {

    public enum CacheType {
        EXPIRATIONS(5 * 60 * 1000L),
        QUOTE(60 * 1000L),
        CHAIN(90 * 1000L);

        final long ttlMs;
        CacheType(long ttlMs) { this.ttlMs = ttlMs; }
    }

    public record CacheEntry<T>(T data, String retrievedAt, long expiresAt) {}

    private final ConcurrentHashMap<String, CacheEntry<?>> store = new ConcurrentHashMap<>();

    /**
     * Get cached response. Returns null on miss or expiry.
     */
    @SuppressWarnings("unchecked")
    public <T> CacheEntry<T> get(CacheType type, String key) {
        String fullKey = type.name().toLowerCase() + ":" + key;
        CacheEntry<?> entry = store.get(fullKey);
        if (entry == null) return null;
        if (System.currentTimeMillis() > entry.expiresAt()) {
            store.remove(fullKey);
            return null;
        }
        return (CacheEntry<T>) entry;
    }

    /**
     * Store a response with type-appropriate TTL.
     */
    public <T> void set(CacheType type, String key, T data, String retrievedAt) {
        String fullKey = type.name().toLowerCase() + ":" + key;
        long expiresAt = System.currentTimeMillis() + type.ttlMs;
        store.put(fullKey, new CacheEntry<>(data, retrievedAt, expiresAt));
    }

    /**
     * Cache statistics for diagnostics.
     */
    public CacheStats stats() {
        int expirations = 0, quote = 0, chain = 0;
        for (String key : store.keySet()) {
            if (key.startsWith("expirations:")) expirations++;
            else if (key.startsWith("quote:")) quote++;
            else if (key.startsWith("chain:")) chain++;
        }
        return new CacheStats(store.size(), expirations, quote, chain);
    }

    public record CacheStats(int size, int expirations, int quote, int chain) {}
}
